File: runround.c
/*
ID: abacles1
LANG: C
TASK: runround
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int inc(char n[])
{
  sprintf(n,"%ld",atol(n)+1);
  return strlen(n);
}

int main()
{
  FILE *in,*out;
  char num[10];
  int digs[9];
  int *touched;
  int rr,di,i,l;
  in = fopen("runround.in","r");
  fscanf(in,"%s",num);
  fclose(in);
  touched = malloc(sizeof(int)*strlen(num));

  while(1)
    {
      for(i=0;i<9;i++)
	digs[i] = 0;
      l = inc(num);
      rr = 1;
      for(i=0;num[i];i++)
	{
	  if(num[i] == '0' || digs[num[i]-'1'] > 0)
	    {
	      rr = 0;
	      break;
	    }
	  digs[num[i]-'1'] = 1;
	}
      if(!rr)
	continue;
      di = 0;
      for(i=0;i<l;i++)
	touched[i] = 0;
      for(i=0;i<l;i++)
	{
	  if(touched[di])
	    {
	      rr = 0;
	      break;
	    }
	  touched[di] = 1;
	  di = (di+(num[di]-'1')+1) % l;
	}
      if(rr && di == 0)
	{
	  out = fopen("runround.out","w");
	  fprintf(out,"%s\n",num);
	  fclose(out);
	  break;
	}
    }

  free(touched);
  return 0;
}

File: runround.py
'''
ID: abacles1
LANG: PYTHON3
TASK: runround
'''

fin = open('runround.in','r')
num = fin.readline().strip()
fin.close()

while True:
    num = str(int(num)+1)
    digs = [0,0,0,0,0,0,0,0,0]
    rr = True
    l = len(num)
    for d in num:
        if d == '0' or digs[ord(d)-ord('1')] > 0:
            rr = False
            break
        digs[ord(d)-ord('1')] += 1
    if not rr:
        continue
    di = 0
    touched = [False]*l
    for i in range(l):
        if touched[di]:
            rr = False
            break
        touched[di] = True
        di = (di+ord(num[di])-ord('1')+1)%l
    if rr and di == 0:
        fout = open('runround.out','w')
        fout.write(num + '\n')
        fout.close()
        print(num)
        break
